enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Enable OCP users to leverage cheaper, non-guaranteed instances to back Machine API Machines.
+
+## Motivation
+
+Allow users to cut costs of running OCP clusters on cloud providers by moving interruptible workloads onto non-guaranteed instances.
+
+### Goals
+
+- Provide sophisticated provider-specific automation for running Machines on non-guaranteed instances
+
+- Utilise as much of the existing Machine API as possible
+
+- Ensure graceful shutdown of pods is attempted on non-guaranteed instances
+
+### Non-Goals
+
+- Any logic for choosing instances types based on availability from the cloud provider
+- A one to one map for each provider available mechanism for deploying spot instances, e.g aws fleet.
+
+## Proposal
+
+To provide a consistent behaviour using non-guaranteed instances (Spot on AWS and Azure, Preepmtible on GCP)
+across cloud providers, we must define a common behaviour based on the common features across each provider.
+
+Based on the research on [non-guaranteed instances](#non-guaranteed-instances),
+the following requirements for integration will work for each of AWS, Azure and GCP:
+
+- Required configuration for enabling spot/preemptible instances should be added to the ProviderSpec
+  - No configuration should be required outside of this scope
+  - This enforces consistency across MachineSets, all Machines a MachinSet creates will either be Spot/Preemptible or On-Demand
+
+- A Machine should be paired 1:1 with an instance on the cloud provider
+  - If the instance is preempted/terminated, the cloud actuator should not replace it
+  - If the instance is preempted/terminated, the cloud provider should not replace it
+
+- The actuator is responsible for creation of the instance only and should not attempt to remediate problems
+
+- The actuator should not attempt to verify that an instance can be created before attempting to create the instance
+  - If the cloud provider does not have capacity, the Machine Health Checker can (given required MHC) remove the Machine after a period.
+    MachineSet will ensure the correct number of Machines are created.
+
+- There’s a [2 hours time frame](https://github.com/openshift/cluster-machine-approver/blob/master/csr_check.go#L38) in OpenShift for a machine to become a node.
+  If the Spot request is not satisfied during this time frame, it is assumed that the instance will never become a node.
+  This machine would be remediated by a given MHC.
+
+
+### Implementation Details
+
+#### Cloud Provider Implementation Specifics
+
+##### AWS
+
+###### Launching instances
+
+To launch an instance as a Spot instance on AWS, a [SpotMarketOptions](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#SpotMarketOptions)
```

Comment:
Here's some existing code to build the options that are more or less outlined here.

https://github.com/openshift/cluster-api-provider-aws/pull/255/files#diff-24a7c87e11cc6413298ddaaa7e7abda0R322